fabtests/benchmarks/rdm_rpc_bw.c

```diff
@@ -0,0 +1,658 @@
+{
+	int ret;
+
+	/* Echo the cookie back to the client. */
+	desc->msg.resp.op = RPC_RESP;
+	desc->msg.resp.status = 0;
+
+	do {
+		ret = fi_tsend(ep, &desc->msg.resp, sizeof(desc->msg.resp),
+			       desc->msg_mr_desc, remote_fi_addr,
+			       desc->msg.req.tag, desc);
+		if (ret == -FI_EAGAIN)
+			fi_cq_read(txcq, NULL, 0);
+	} while (ret == -FI_EAGAIN);
+
+	if (ret)
+		FT_ERR("fi_send failed: %d", ret);
+
+	return ret;
+}
+
+static int rpc_desc_server_process_send_response(struct rpc_desc *desc)
+{
+	/* Server has completed the RPC. Metrics can be updated. */
+	rpc_byte_count += desc->msg.req.len;
+	rpc_count++;
+	rpc_inflight--;
+
+	return 0;
+}
+
+static int rpc_desc_server_post_recv(struct rpc_desc *desc)
+{
+	int ret;
+
+	/* Reuse the RPC descriptor for the next RPC request. Unlike the client,
+	 *no cleanup and re-init of the RPC descriptor happens.
+	 */
+	do {
+		ret = fi_recv(ep, &desc->msg.req, sizeof(desc->msg.req),
+			      desc->msg_mr_desc, FI_ADDR_UNSPEC, desc);
+		if (ret == -FI_EAGAIN)
+			fi_cq_read(rxcq, NULL, 0);
+	} while (ret == -FI_EAGAIN);
+
+	if (ret)
+		FT_ERR("fi_recv failed: %d", ret);
+	else
+		rpc_inflight++;
+
+	return ret;
+}
+
+static int rpc_desc_client_send_req(struct rpc_desc *desc)
+{
+	int ret;
+
+	/* Register the payload buffer and prepare the RPC request. The payload
+	 * buffer should have already been allocated during RPC descriptor init.
+	 */
+	ret = rpc_desc_reg_payload_mr(desc);
+	if (ret) {
+		FT_ERR("fi_trecv failed: %d", ret);
+		goto err;
+	}
+
+	/* Prepost tagged receive buffer to handle response. */
+	do {
+		ret = fi_trecv(ep, &desc->msg.resp, sizeof(desc->msg.resp),
+			       desc->msg_mr_desc, FI_ADDR_UNSPEC, desc->tag, 0,
+			       desc);
+		if (ret == -FI_EAGAIN)
+			fi_cq_read(rxcq, NULL, 0);
+	} while (ret == -FI_EAGAIN);
+
+	if (ret) {
+		FT_ERR("fi_trecv failed: %d", ret);
+		goto err_dereg_mr;
+	}
+
+	/* Initialize the RPC request message and send it to the server.
+	 * Tag is defined during RPC descriptor initialization.
+	 * rpc_desc_reg_payload_mr() fills in the remaining fields.
+	 */
+	rpc_desc_client_init_req(desc);
+
+	do {
+		ret = fi_send(ep, &desc->msg.req, sizeof(desc->msg.req),
+			      desc->msg_mr_desc, remote_fi_addr, desc);
+		if (ret == -FI_EAGAIN)
+			fi_cq_read(txcq, NULL, 0);
+	} while (ret == -FI_EAGAIN);
+
+	if (ret) {
+		FT_ERR("fi_send failed: %d", ret);
+		goto err_cancel_trecv;
+	} else {
+		rpc_inflight++;
+	}
+
+	return 0;
+
+err_cancel_trecv:
+	fi_cancel(&ep->fid, desc);
+err_dereg_mr:
+	FT_CLOSE_FID(desc->payload_mr);
+err:
+	return ret;
+}
+
+static int rpc_desc_client_process_recv(struct rpc_desc *desc)
+{
+	if (desc->msg.resp.status != 0) {
+		FT_ERR("Bad server response status");
+		return -FI_EIO;
+	}
+
+	/* Server has completed the RPC. Metrics can be updated. */
+	rpc_byte_count += desc->msg.req.len;
+	rpc_count++;
+	rpc_inflight--;
+
+	FT_CLOSE_FID(desc->payload_mr);
+
+	return 0;
+}
+
+static void rpc_process_tx_cq(size_t rpc_limit)
```

Comment:
I found the term rpc_limit to be confusing, particularly when it's frequented compared alongside the window size.  I *think* this is the number of iterations to execute, rather than some 'limit'.